File: demos/bitcast/bench-decode.go
package main

import (
	"fmt"
	"os"
	"time"
	"unsafe"
)

type Uint128 struct {
	data [16]byte
}

type Transfer struct {
	id                Uint128
	debit_account_id  Uint128
	credit_account_id Uint128
	user_data         Uint128
	reserved          Uint128
	pending_iD        Uint128
	timeout           uint64
	ledger            uint32
	code              uint16
	flags             uint16
	amount            uint64
	timestamp         uint64
}

const TRANSFER int = 128

func main() {

	buffer := load("transfers")

	var loops int = 10
	for loops > 0 {
		var offset int = 0
		var sum uint64 = 0
		now := time.Now()

		for offset < len(buffer) {

			// Deserialize without any overhead:
			transfer := (*Transfer)(unsafe.Pointer(&buffer[offset]))
			sum += transfer.amount
			offset += TRANSFER
		}

		elapsed := time.Since(now)
		fmt.Printf("  go: sum of transfer amounts=%d ns=%d\n", sum, elapsed.Nanoseconds())
		loops -= 1
	}
}

func load(name string) []byte {

	file, openErr := os.Open(name)
	if openErr != nil {
		panic(openErr)
	}
	defer file.Close()

	stats, statsErr := file.Stat()
	if statsErr != nil {
		panic(statsErr)
	}

	var size int64 = stats.Size()
	bytes := make([]byte, size)

	_, readErr := file.Read(bytes)
	if readErr != nil {
		panic(readErr)
	}

	return bytes
}

File: demos/bitcast/bench-decode.cs
using System;
using System.Runtime.InteropServices;
public static class BenchDecode
{
    const int TRANSFER = 128;

    [StructLayout(LayoutKind.Sequential, Size = 16)]
    unsafe struct UInt128
    {
        public fixed byte bytes[16];
    }    

    [StructLayout(LayoutKind.Sequential, Size = TRANSFER)]
    unsafe struct Transfer
    {
        public UInt128 id;
        public UInt128 debit_account_id;
        public UInt128 credit_account_id;
        public UInt128 user_data;
        public UInt128 reserved;
        public UInt128 pending_id;
        public ulong timeout;
        public uint ledger;
        public ushort code;
        public ushort flags;
        public ulong amount;
        public ulong timestamp;
    }

    public static void Main() {

        var buffer = Load("transfers");

        Console.WriteLine("do this a few times to let CLR optimize...");

        int loops = 10;
        while (loops-- > 0)
        {
            int offset = 0;
            ulong sum = 0UL;
            var now = DateTime.Now;

            while (offset < buffer.Length) 
            {
                var array = buffer[offset..][0..TRANSFER];
                
                // Deserialize without much overhead:
                var transfer = MemoryMarshal.AsRef<Transfer>(array);

                sum += transfer.amount;
                offset += TRANSFER;
            }

            TimeSpan elapsed = DateTime.Now - now;
            Console.WriteLine("  C#: sum of transfer amounts={0} ms={1:0.000}", sum, elapsed.TotalMilliseconds);
        }
    }

    static Span<byte> Load(string file)
    {
        return System.IO.File.ReadAllBytes(file).AsSpan(0);
    }



}